### ISSUE TITLE
Panics while loading assets kill IoThreads preventing assets to be further loaded

### ISSUE THREAD

**Issue by lassade** [bevyengine/bevy#2302]
## Bevy version
main branch c4b8210

## Operating system & version
Fedora 33

## What you did
Write any AssetLoader that panics, like so:
```rust
#[derive(Default)]
pub struct EvilLoader;

impl AssetLoader for EvilLoader {
    fn load<'a>(
        &'a self,
        bytes: &'a [u8],
        load_context: &'a mut LoadContext,
    ) -> BoxedFuture<'a, Result<()>> {
        Box::pin(async move { panic!("malevolence") })
    }

    fn extensions(&self) -> &[&str] {
        &["evil"]
    }
}
```
Any attempt to load the evil asset will kill the IO Thread, in a 4 thread cpu like mine this means that if I got 1 panic not asset can be loaded after

## What you expected to happen
A panic should closes bevy or at least restart the IO Thread so the app doesn't hangs

## What actually happened
Bevy just hangs and nothing else loads

**Comment by mockersf** [comment]
> A panic should closes bevy or at least restart the IO Thread so the app doesn't hangs

I can see the case for both but I would prefer it to restart the dead threads. Ideally it could be configured?

**Comment by NathanSWard** [comment]
This will be a little tricky to implement due to the restrictions of `UnwindSafe`. 
however, we definitely need a way to detect the panic and the restart the thread or bubble the panic up to the bevy main thread.

**Comment by mockersf** [comment]
fixed by #6443
